FirebaseAdmin/FirebaseAdmin/Auth/AbstractFirebaseAuth.cs

```diff
@@ -0,0 +1,587 @@
+        /// serialization it should not be larger than 1000 characters.</param>
+        /// <param name="cancellationToken">A cancellation token to monitor the asynchronous
+        /// operation.</param>
+        public async Task SetCustomUserClaimsAsync(
+            string uid, IReadOnlyDictionary<string, object> claims, CancellationToken cancellationToken)
+        {
+            var userManager = this.IfNotDeleted(() => this.userManager.Value);
+            var user = new UserRecordArgs()
+            {
+                Uid = uid,
+                CustomClaims = claims,
+            };
+
+            await userManager.UpdateUserAsync(user, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets an async enumerable to iterate or page through users starting from the specified
+        /// page token. If the page token is null or unspecified, iteration starts from the first
+        /// page. See <a href="https://googleapis.github.io/google-cloud-dotnet/docs/guides/page-streaming.html">
+        /// Page Streaming</a> for more details on how to use this API.
+        /// </summary>
+        /// <param name="options">The options to control the starting point and page size. Pass null
+        /// to list from the beginning with default settings.</param>
+        /// <returns>A <see cref="PagedAsyncEnumerable{ExportedUserRecords, ExportedUserRecord}"/> instance.</returns>
+        public PagedAsyncEnumerable<ExportedUserRecords, ExportedUserRecord> ListUsersAsync(
+            ListUsersOptions options)
+        {
+            var userManager = this.IfNotDeleted(() => this.userManager.Value);
+
+            return userManager.ListUsers(options);
+        }
+
+        /// <summary>
+        /// Deletes this service instance.
+        /// </summary>
+        void IFirebaseService.Delete()
+        {
+            lock (this.authLock)
+            {
+                this.deleted = true;
+                this.tokenFactory.DisposeIfCreated();
+                this.userManager.DisposeIfCreated();
+            }
+        }
+
+        private TResult IfNotDeleted<TResult>(Func<TResult> func)
+        {
+            lock (this.authLock)
+            {
+                if (this.deleted)
+                {
+                    throw new InvalidOperationException("Cannot invoke after deleting the app.");
+                }
+
+                return func();
+            }
+        }
+
+        internal sealed class FirebaseAuthArgs
+        {
+            internal Lazy<FirebaseTokenFactory> TokenFactory { get; set; }
+
+            internal Lazy<FirebaseTokenVerifier> IdTokenVerifier { get; set; }
+
+            internal Lazy<FirebaseUserManager> UserManager { get; set; }
+
+            internal Lazy<TenantManager> TenantManager { get; set; }
+
+            internal string TenantId { get; set; }
+
+            internal static FirebaseAuthArgs Create(FirebaseApp app, string tenantId)
```

Comment:
Logic for creating `FirebaseAuthArgs` should go into `FirebaseAuth` and `TenantAwareFirebaseAuth` classes.